File: synopsis/static/cloud.js
// Cloud storage controls share the existing settings and document panes.
export function connectorFields(config, escape, safeUrl) {
  return `<section class="settings-section" id="cloud-connectors"><div class="settings-heading">Cloud document storage</div>
    <p class="field-hint">Save new uploads and watched-folder imports to a connected drive. A local copy stays in your upload folder for OCR, reading, and backups. Existing documents can be saved from their details pane.</p>
    <label class="form-field">Save new documents to<select id="cloud-destination" aria-label="Save new documents to"><option value="local" ${config.destination==='local'?'selected':''}>Local folder only</option>${Object.entries(config.providers).map(([id,c])=>`<option value="${id}" ${config.destination===id?'selected':''} ${!c.connected?'disabled':''}>${escape(c.name)}${!c.connected?' (connect first)':''}</option>`).join('')}</select></label>
    <button type="button" class="button secondary small" data-cloud-action="destination">Apply storage destination</button>
    ${Object.entries(config.providers).map(([id,c])=>`<details class="connector-card" data-connector="${id}" ${c.connected?'':'open'}><summary><strong>${escape(c.name)}</strong><span>${c.connected?'Connected':'Not connected'}</span></summary>
      ${c.connected?`<p class="field-hint">${escape(c.account)} · ${escape(c.folderName)}</p>${c.folderUrl?`<a class="text-button" href="${escape(safeUrl(c.folderUrl))}" target="_blank" rel="noopener noreferrer">Open cloud folder</a>`:''}<button type="button" class="text-button danger" data-cloud-action="disconnect" data-provider="${id}">Disconnect ${escape(c.name)}</button>`:`
      <p class="field-hint">Register an OAuth web application with ${id==='google'?'Google Cloud and enable the Drive API':'Microsoft Entra (personal and/or work accounts)'}. Add this exact redirect URI:</p>
      <code class="connector-redirect">${escape(c.redirectUri)}</code>
      <p class="field-hint">${id==='google'?'Use the drive.file scope. Synopsis creates a dedicated folder when you connect.':'Use delegated Files.ReadWrite and offline_access permissions. The folder is created in your OneDrive if needed. If the portal rejects an HTTP 127.0.0.1 redirect, reopen Synopsis using localhost with the same port, then copy the redirect URI shown there.'} <a href="${id==='google'?'https://developers.google.com/identity/protocols/oauth2/web-server':'https://learn.microsoft.com/en-us/entra/identity-platform/quickstart-register-app'}" target="_blank" rel="noopener noreferrer">Provider setup guide</a></p>
      <label class="form-field">${escape(c.name)} client ID<input data-cloud-field="clientId" value="${escape(c.clientId)}" maxlength="500" autocomplete="off"></label>
      <label class="form-field">${escape(c.name)} client secret<input data-cloud-field="clientSecret" type="password" maxlength="4096" autocomplete="new-password" placeholder="${c.hasClientSecret?'Leave blank to keep saved secret':'OAuth application secret'}"></label>
      ${id==='onedrive'?`<label class="form-field">Microsoft tenant<input data-cloud-field="tenant" value="${escape(c.tenant||'common')}" placeholder="common" maxlength="100"></label>`:''}
      <label class="form-field">${escape(c.name)} folder name<input data-cloud-field="folderName" value="${escape(c.folderName||'Synopsis')}" maxlength="100"></label>
      <button type="button" class="button secondary small" data-cloud-action="connect" data-provider="${id}">Connect ${escape(c.name)}</button>`}
    </details>`).join('')}
    <p class="field-hint">Connecting opens provider sign-in in this tab; save other preference edits first. Connection changes and Apply storage destination save immediately. Credentials stay in the local database and are excluded from exported backups. Disconnecting or deleting a local reference keeps its cloud copy.</p>
    <p id="cloud-settings-status" role="status" class="field-hint"></p>
  </section>`;
}

export function cloudDetail(item, escape, safeUrl) {
  if(!item.fileName)return '';
  const job=item.cloudStorage, name=job?.provider==='google'?'Google Drive':'OneDrive';
  return `<div class="cloud-document"><div class="detail-section-title">CLOUD STORAGE</div>${job?`<p class="field-hint"><strong>${escape(name)} · ${escape(({queued:'Waiting to upload',uploading:'Uploading…',saved:'Saved',error:'Needs attention',paused:'Paused'})[job.status]||job.status)}</strong></p>${job.error?`<p class="field-hint cloud-transfer-error">${escape(job.error)}</p>`:''}${job.url?`<a class="text-button" href="${escape(safeUrl(job.url))}" target="_blank" rel="noopener noreferrer">Open cloud copy</a>`:''}`:'<p class="field-hint">Local copy. Choose a connected drive in Settings to save this document to cloud.</p>'}
    ${!item.trashed?`<div class="field-buttons">${job?.status==='error'?'<button type="button" class="text-button" data-cloud-action="retry">Retry cloud upload</button>':''}<button type="button" class="text-button" data-cloud-action="save" ${['queued','uploading'].includes(job?.status)?'disabled':''}>Save to cloud</button></div>`:''}</div>`;
}

export function installCloudUI({api, escape, safeUrl, toast, refresh, state, openSettings}) {
  document.addEventListener('click',async event=>{
    const button=event.target.closest('[data-cloud-action]');if(!button)return;
    const action=button.dataset.cloudAction, provider=button.dataset.provider;
    button.disabled=true;
    const status=document.querySelector('#cloud-settings-status');
    try {
      if(action==='connect') {
        const card=button.closest('[data-connector]'), body={};
        card.querySelectorAll('[data-cloud-field]').forEach(input=>body[input.dataset.cloudField]=input.value);
        await api(`/connectors/${provider}/configure`,{method:'POST',body});
        card.querySelector('[type=password]').value='';
        const result=await api(`/connectors/${provider}/connect`,{method:'POST'});
        location.assign(result.url);return;
      }
      if(action==='disconnect') {
        if(!confirm('Disconnect this account? Future uploads will use local storage. Existing cloud copies remain in your drive.'))return;
        await api(`/connectors/${provider}/disconnect`,{method:'POST'});
      }else if(action==='destination') {
        await api('/connectors/destination',{method:'POST',body:{destination:document.querySelector('#cloud-destination').value}});
      }else if(action==='save'||action==='retry') {
        await api(`/connectors/items/${state.selected}/${action}`,{method:'POST'});
        await refresh({detail:true});toast('Cloud transfer queued');return;
      }
      const config=await api('/connectors');
      document.querySelector('#cloud-connectors').outerHTML=connectorFields(config,escape,safeUrl);
      document.querySelector('#cloud-settings-status').textContent=action==='disconnect'?'Account disconnected. Cloud copies were kept.':'Storage destination saved.';
      await refresh({detail:true});
    }catch(error){if(status)status.textContent=error.message;toast(error.message,true);}
    finally{button.disabled=false;}
  });
  const result=new URL(location.href).searchParams.get('connector');
  if(result){history.replaceState(null,'',location.pathname);openSettings('cloud').then(()=>{
    document.querySelector('#cloud-connectors')?.scrollIntoView();
    document.querySelector('#cloud-settings-status').textContent=result==='connected'?'Account connected. Select it as the storage destination to save new uploads.':'Connection failed or was canceled. Check your client credentials, exact redirect URI, consent permissions, and network, then connect again.';
  }).catch(error=>toast(error.message,true));}
}

File: synopsis/cloud.py
"""OAuth-connected cloud copies of originals; local files remain the OCR source."""
import base64
import hashlib
import hmac
import mimetypes
import re
import secrets
import time
from functools import wraps
from threading import Event, RLock, Thread
from urllib.parse import quote, urlencode, urlsplit
from uuid import uuid4

import requests
from flask import Blueprint, abort, jsonify, redirect, request
from werkzeug.utils import secure_filename

from .storage import now

PROVIDERS = {'google': 'Google Drive', 'onedrive': 'OneDrive'}
GOOGLE = 'https://www.googleapis.com/drive/v3'
GRAPH = 'https://graph.microsoft.com/v1.0'
SCOPES = {'google': 'https://www.googleapis.com/auth/drive.file',
          'onedrive': 'offline_access https://graph.microsoft.com/Files.ReadWrite'}


class CloudError(ValueError):
    pass


def http(method, url, **kwargs):
    try:
        response = requests.request(method, url, timeout=(10, 120), allow_redirects=False, **kwargs)
    except requests.RequestException:
        raise CloudError('Cloud service unavailable. Your local document is safe; retry the transfer.') from None
    if response.status_code not in (200, 201, 204, 404, 409):
        if response.status_code in (401, 403):
            raise CloudError('Cloud access was denied. Check account permissions or reconnect in Settings.')
        if response.status_code == 429:
            raise CloudError('The cloud service is busy. Wait a moment, then retry.')
        raise CloudError('Cloud request failed. Check account storage and connectivity, then retry.')
    return response


def data(response):
    if response.status_code not in (200, 201):
        raise CloudError('Cloud folder or file is unavailable. Check the connection in Settings.')
    try:
        result = response.json()
        if not isinstance(result, dict): raise ValueError()
        return result
    except (ValueError, TypeError):
        raise CloudError('The cloud provider returned an invalid response.') from None


class CloudStorage:
    def __init__(self, store):
        self.store = store
        self.lock = RLock()
        self.pending = {}
        self.stop_event = Event()
        self.thread = None

    def config(self, provider):
        return self.store.setting('cloud:' + provider, {})

    def synchronized(self, fn):
        @wraps(fn)
        def wrapped(*args, **kwargs):
            if not self.lock.acquire(blocking=False):
                abort(409, description='A cloud transfer is in progress. Try this action after it finishes.')
            try:
                return fn(*args, **kwargs)
            finally:
                self.lock.release()
        return wrapped

    def save(self, provider, config):
        self.store.set_setting('cloud:' + provider, config)

    def public(self, origin):
        result = {}
        for provider, name in PROVIDERS.items():
            c = self.config(provider)
            result[provider] = {key: c.get(key, '') for key in ('clientId', 'tenant', 'folderName', 'account', 'folderUrl')}
            result[provider].update(name=name, connected=bool(c.get('refreshToken')), hasClientSecret=bool(c.get('clientSecret')),
                                    redirectUri=origin + '/api/connectors/' + provider + '/callback')
        return {'destination': self.store.setting('cloudDestination', 'local'), 'providers': result}

    def configure(self, provider, values):
        c = self.config(provider)
        if c.get('refreshToken'):
            raise ValueError('Disconnect this account before changing its application settings.')
        fields = {}
        for key, default, limit in (('clientId', '', 500), ('clientSecret', '', 4096), ('tenant', 'common', 100), ('folderName', 'Synopsis', 100)):
            value = values.get(key, default)
            if not isinstance(value, str) or len(value) > limit or any(ord(v) < 32 for v in value):
                raise ValueError('Enter valid connector settings.')
            fields[key] = value.strip()
        if not fields['clientId']: raise ValueError('An OAuth application client ID is required.')
        if not re.fullmatch(r'[A-Za-z0-9.-]+', fields['tenant']): raise ValueError('Enter a valid Microsoft tenant ID or common.')
        if not fields['folderName'] or re.search(r'["*:<>?/\\|#%]', fields['folderName']) or fields['folderName'].endswith('.'):
            raise ValueError('Use a simple folder name without path separators or reserved characters.')
        if not fields['clientSecret'] and fields['clientId'] == c.get('clientId') and fields['tenant'] == c.get('tenant'):
            fields['clientSecret'] = c.get('clientSecret', '')
        if not fields['clientSecret']: raise ValueError('An OAuth web application client secret is required.')
        self.save(provider, fields)
        self.pending.pop(provider, None)

    def token_url(self, provider, c):
        return 'https://oauth2.googleapis.com/token' if provider == 'google' else f'https://login.microsoftonline.com/{c["tenant"]}/oauth2/v2.0/token'

    def begin(self, provider, origin):
        c = self.config(provider)
        if not c.get('clientId') or not c.get('clientSecret'): raise ValueError('Save OAuth application settings before connecting.')
        state, verifier, binding = secrets.token_urlsafe(32), secrets.token_urlsafe(48), secrets.token_urlsafe(32)
        callback = origin + '/api/connectors/' + provider + '/callback'
        self.pending[provider] = {'state': state, 'verifier': verifier, 'binding': binding, 'expires': time.time()+600, 'callback': callback}
        params = {'client_id': c['clientId'], 'redirect_uri': callback, 'response_type': 'code', 'scope': SCOPES[provider],
                  'state': state, 'code_challenge': base64.urlsafe_b64encode(hashlib.sha256(verifier.encode()).digest()).decode().rstrip('='),
                  'code_challenge_method': 'S256'}
        if provider == 'google':
            url = 'https://accounts.google.com/o/oauth2/v2/auth'
            params.update(access_type='offline', prompt='consent')
        else:
            url = f'https://login.microsoftonline.com/{c["tenant"]}/oauth2/v2.0/authorize'
            params.update(response_mode='query', prompt='select_account')
        return url + '?' + urlencode(params), binding

    def token(self, provider, c, fields):
        payload = {'client_id': c['clientId'], 'client_secret': c['clientSecret'], **fields}
        if provider == 'onedrive': payload['scope'] = SCOPES[provider]
        value = data(http('POST', self.token_url(provider, c), data=payload))
        if not isinstance(value.get('access_token'), str) or not value['access_token']:
            raise CloudError('No cloud access token received. Reconnect the account.')
        refresh = value.get('refresh_token') or c.get('refreshToken')
        if not isinstance(refresh, str) or not refresh:
            raise CloudError('Offline access was not granted. Reconnect and allow offline file access.')
        return {**c, 'accessToken': value['access_token'], 'refreshToken': refresh,
                'expiresAt': time.time() + max(0, int(value.get('expires_in', 3600)))}

    def finish(self, provider, args, binding):
        pending = self.pending.get(provider)
        if not pending or pending['expires'] < time.time() or not hmac.compare_digest(args.get('state', ''), pending['state']) or not hmac.compare_digest(binding, pending['binding']):
            raise ValueError('This connection request expired or belongs to another browser. Start again in Settings.')
        self.pending.pop(provider)
        if args.get('error'): raise ValueError('Sign-in was canceled or access was denied. Try connecting again.')
        code = args.get('code', '')
        if not code or len(code) > 8192: raise ValueError('No valid authorization code was received.')
        c = self.config(provider)
        if c.get('refreshToken'): raise ValueError('Disconnect the current account before connecting another.')
        c = self.token(provider, c, {'grant_type': 'authorization_code', 'code': code,
                                    'redirect_uri': pending['callback'], 'code_verifier': pending['verifier']})
        headers = {'Authorization': 'Bearer ' + c['accessToken']}
        if provider == 'google':
            user = data(http('GET', GOOGLE+'/about', headers=headers, params={'fields': 'user'})).get('user', {})
            c.update(account=user.get('emailAddress') or user.get('displayName') or 'Google Drive', accountId=user.get('permissionId', ''))
            folder = data(http('POST', GOOGLE+'/files', headers=headers,
                               json={'name': c['folderName'], 'mimeType': 'application/vnd.google-apps.folder'},
                               params={'fields': 'id,webViewLink'}))
            c['folderUrl'] = folder.get('webViewLink', '')
        else:
            drive = data(http('GET', GRAPH+'/me/drive', headers=headers, params={'$select': 'id,owner'}))
            user = drive.get('owner', {}).get('user', {})
            c.update(account=user.get('email') or user.get('displayName') or 'OneDrive', accountId=drive['id'])
            folder_url = GRAPH+'/me/drive/root:/'+quote(c['folderName'], safe='')
            response = http('GET', folder_url, headers=headers)
            if response.status_code == 404:
                response = http('POST', GRAPH+'/me/drive/root/children', headers=headers,
                                json={'name': c['folderName'], 'folder': {}, '@microsoft.graph.conflictBehavior': 'fail'})
                if response.status_code == 409: response = http('GET', folder_url, headers=headers)
            folder = data(response)
            if 'folder' not in folder: raise ValueError('The chosen OneDrive folder name belongs to a file. Choose another name.')
            c['folderUrl'] = folder.get('webUrl', '')
        if not isinstance(folder.get('id'), str) or not folder['id']: raise CloudError('No cloud folder was returned.')
        c.update(folderId=folder['id'], connectionId=str(uuid4()))
        self.save(provider, c)

    def access(self, provider, c):
        if not c.get('refreshToken'): raise CloudError('This cloud account is disconnected. Connect it in Settings.')
        if c.get('expiresAt', 0) < time.time()+60:
            c = self.token(provider, c, {'grant_type': 'refresh_token', 'refresh_token': c['refreshToken']})
            self.save(provider, c)
        return c

    def disconnect(self, provider):
        c = self.config(provider)
        self.save(provider, {k: c.get(k, '') for k in ('clientId', 'clientSecret', 'tenant', 'folderName')})
        self.pending.pop(provider, None)
        if self.store.setting('cloudDestination') == provider: self.store.set_setting('cloudDestination', 'local')
        for item in self.store.items():
            job = item.get('cloudStorage', {})
            if job.get('provider') == provider and job.get('status') in ('queued', 'uploading'):
                self.store.update(item['id'], {'cloudStorage': {**job, 'status': 'error', 'error': 'Account disconnected. Connect and save this document to cloud again.'}})

    def destination(self, value):
        if value not in ('local', *PROVIDERS): raise ValueError('Choose local storage, Google Drive, or OneDrive.')
        if value != 'local' and not self.config(value).get('refreshToken'): raise ValueError('Connect the cloud account before selecting it.')
        self.store.set_setting('cloudDestination', value)

    def enqueue(self, item_id, explicit=False):
        # Snapshot credentials/destination without waiting for a network transfer.
        # A disconnected/replaced connection is rejected again by the worker.
        item = self.store.get(item_id)
        if not item or not item.get('filePath') or item.get('trashed'): return
        old = item.get('cloudStorage', {})
        if old and not explicit: return
        if old.get('status') == 'uploading': raise ValueError('This cloud transfer is already in progress.')
        provider = self.store.setting('cloudDestination', 'local')
        if provider == 'local':
            if explicit: raise ValueError('Choose a cloud destination in Settings first.')
            return
        c = self.config(provider)
        if not c.get('refreshToken'):
            if explicit: raise ValueError('Connect the cloud account in Settings first.')
            self.store.update(item_id, {'cloudStorage': {'provider': provider, 'connectionId': '', 'status': 'error', 'error': 'Account disconnected during intake. Connect and use Save to cloud.'}})
            return
        if old.get('provider') == provider and old.get('connectionId') == c['connectionId']:
            if old.get('status') == 'saved': return
            job = {**old, 'status': 'queued', 'error': ''}
        else:
            job = {'provider': provider, 'connectionId': c['connectionId'], 'folderId': c['folderId'], 'status': 'queued', 'error': ''}
        self.store.update(item_id, {'cloudStorage': job})

    def retry(self, item_id):
        item = self.store.get(item_id)
        job = item.get('cloudStorage', {}) if item else {}
        if not job or job.get('status') != 'error': raise ValueError('There is no failed cloud transfer to retry.')
        if job['connectionId'] != self.config(job['provider']).get('connectionId'):
            raise ValueError('The account connection changed. Use Save to cloud to choose the current destination.')
        self.store.update(item_id, {'cloudStorage': {**job, 'status': 'queued', 'error': ''}})

    def transfer(self, item_id):
        with self.lock:
            item = self.store.get(item_id)
            job = item.get('cloudStorage', {}) if item else {}
            if not job or job.get('status') not in ('queued', 'uploading'): return
            if item.get('trashed'):
                self.store.update(item_id, {'cloudStorage': {**job, 'status': 'error', 'error': 'Document is in Trash. Restore it before retrying.'}})
                return
            job = {**job, 'status': 'uploading', 'error': ''}
            self.store.update(item_id, {'cloudStorage': job})
            try:
                provider = job['provider']
                c = self.config(provider)
                if job['connectionId'] != c.get('connectionId'): raise CloudError('Account connection changed. Use Save to cloud to choose the current destination.')
                c = self.access(provider, c)
                path = self.store.document_path(item)
                size = path.stat().st_size
                if not 0 < size <= 100*1024*1024: raise CloudError('Cloud documents must be between 1 byte and 100 MB.')
                headers = {'Authorization': 'Bearer ' + c['accessToken']}
                mime = mimetypes.guess_type(path.name)[0] or 'application/octet-stream'
                if provider == 'google':
                    if not job.get('remoteId'):
                        job['remoteId'] = data(http('GET', GOOGLE+'/files/generateIds', headers=headers, params={'count': 1, 'space': 'drive'}))['ids'][0]
                        self.store.update(item_id, {'cloudStorage': job})
                    remote = GOOGLE+'/files/'+quote(job['remoteId'], safe='')
                    fields = {'fields': 'id,size,md5Checksum,webViewLink,trashed'}
                    response = http('GET', remote, headers=headers, params=fields)
                    if response.status_code == 404:
                        response = http('POST', 'https://www.googleapis.com/upload/drive/v3/files', headers={**headers, 'X-Upload-Content-Type': mime, 'X-Upload-Content-Length': str(size)},
                                        params={'uploadType': 'resumable', **fields}, json={'id': job['remoteId'], 'name': item['fileName'], 'parents': [job['folderId']]})
                        if response.status_code == 409:
                            response = http('GET', remote, headers=headers, params=fields)
                        else:
                            location = response.headers.get('Location', '')
                            parsed = urlsplit(location)
                            if parsed.scheme != 'https' or parsed.netloc != 'www.googleapis.com' or not parsed.path.startswith('/upload/'):
                                raise CloudError('Invalid Google Drive upload session. Retry the transfer.')
                            with path.open('rb') as stream:
                                response = http('PUT', location, headers={**headers, 'Content-Type': mime, 'Content-Length': str(size)}, data=stream)
                    result = data(response)
                    with path.open('rb') as stream: checksum = hashlib.file_digest(stream, 'md5').hexdigest()
                    if result.get('md5Checksum') != checksum or result.get('trashed'):
                        raise CloudError('The remote document does not match the local original. Check the cloud copy.')
                    url = result.get('webViewLink', '') or 'https://drive.google.com/file/d/'+quote(job['remoteId'], safe='')+'/view'
                else:
                    # Stable unique path makes retries replace the same Synopsis copy.
                    name = item['id']+'-'+(secure_filename(item['fileName']) or path.name)[-140:]
                    remote = GRAPH+'/me/drive/items/'+quote(job['folderId'], safe='')+':/'+quote(name, safe='')+':/content'
                    with path.open('rb') as stream:
                        result = data(http('PUT', remote, headers={**headers, 'Content-Type': mime, 'Content-Length': str(size)}, data=stream))
                    url = result.get('webUrl', '')
                if int(result.get('size', -1)) != size or not result.get('id'): raise CloudError('The cloud upload size could not be verified. Retry the transfer.')
                job.update(status='saved', remoteId=result['id'], url=url, savedAt=now(), error='')
            except (ValueError, OSError, KeyError, TypeError, IndexError) as exc:
                job.update(status='error', error=str(exc) if isinstance(exc, CloudError) else 'Cloud transfer failed. Your local document is safe; retry or reconnect in Settings.')
            self.store.update(item_id, {'cloudStorage': job})

    def start(self):
        def run():
            while not self.stop_event.is_set():
                for item in self.store.items():
                    if self.stop_event.is_set(): break
                    if item.get('cloudStorage', {}).get('status') in ('queued', 'uploading'):
                        self.transfer(item['id'])
                self.stop_event.wait(3)
        self.thread = Thread(target=run, name='synopsis-cloud', daemon=True)
        self.thread.start()

    def close(self):
        self.stop_event.set()


def register_cloud(app, store):
    cloud = CloudStorage(store)
    app.extensions['cloud'] = cloud
    bp = Blueprint('connectors', __name__, url_prefix='/api/connectors')

    @bp.before_request
    def provider_check():
        if request.view_args and request.view_args.get('provider') not in (None, *PROVIDERS): abort(404)

    @bp.get('')
    def settings():
        return jsonify(cloud.public(request.host_url.rstrip('/')))

    @bp.post('/destination')
    @cloud.synchronized
    def destination():
        with cloud.lock: cloud.destination(request.get_json().get('destination'))
        return jsonify(ok=True)

    @bp.post('/<provider>/configure')
    @cloud.synchronized
    def configure(provider):
        with cloud.lock: cloud.configure(provider, request.get_json())
        return jsonify(ok=True)

    @bp.post('/<provider>/connect')
    @cloud.synchronized
    def connect(provider):
        with cloud.lock:
            if cloud.config(provider).get('refreshToken'): raise ValueError('Disconnect this account before connecting again.')
            url, binding = cloud.begin(provider, request.host_url.rstrip('/'))
        response = jsonify(url=url)
        response.set_cookie('synopsis-oauth-'+provider, binding, max_age=600, httponly=True, samesite='Lax', secure=request.is_secure, path='/api/connectors/'+provider)
        return response

    @bp.get('/<provider>/callback')
    def callback(provider):
        try:
            with cloud.lock: cloud.finish(provider, request.args, request.cookies.get('synopsis-oauth-'+provider, ''))
            result = 'connected'
        except (ValueError, KeyError, TypeError):
            result = 'failed'
        response = redirect('/?connector='+result)
        response.delete_cookie('synopsis-oauth-'+provider, path='/api/connectors/'+provider)
        return response

    @bp.post('/<provider>/disconnect')
    @cloud.synchronized
    def disconnect(provider):
        with cloud.lock: cloud.disconnect(provider)
        return jsonify(ok=True)

    @bp.post('/items/<item_id>/save')
    @cloud.synchronized
    def save(item_id):
        item = store.get(item_id)
        if not item or not item.get('filePath') or item.get('trashed'): raise ValueError('Choose an active document to save to cloud.')
        cloud.enqueue(item_id, explicit=True)
        return jsonify(ok=True), 202

    @bp.post('/items/<item_id>/retry')
    @cloud.synchronized
    def retry(item_id):
        with cloud.lock: cloud.retry(item_id)
        return jsonify(ok=True), 202

    app.register_blueprint(bp)
    if app.config.get('CLOUD_JOBS', app.config['PROCESS_JOBS'] and not app.config.get('TESTING')):
        cloud.start()
    return cloud
